api/e2e/test/routers_suite_test.go

```diff
@@ -0,0 +1,271 @@
+	if len(args) == 0 {
+		return false
+	}
+
+	var (
+		body          func(ctx *RouterContext)
+		ok            bool
+		pythonVersion string
+		testCaseName  string
+	)
+	if body, ok = args[len(args)-1].(func(ctx *RouterContext)); !ok {
+		panic("last argument must have type func(ctx *RouterContext)")
+	}
+	if pythonVersion, ok = args[0].(string); ok {
+		args = args[1 : len(args)-1]
+		testCaseName = fmt.Sprintf("python-%s:%s", pythonVersion, payloadTpl)
+	} else {
+		args = args[:len(args)-1]
+		testCaseName = payloadTpl
+	}
+
+	return Context("Turing", append(args, Ordered, func() {
+		var (
+			e         *httpexpect.Expect
+			routerCtx RouterContext
+			router    *httpexpect.Object
+		)
+
+		BeforeAll(func() {
+			e = config.NewHTTPExpect(GinkgoT(), cfg.APIBasePath)
+
+			var ensembler config.EnsemblerData
+			if pythonVersion != "" {
+				for _, e := range cfg.Ensemblers.Entities {
+					if strings.HasPrefix(e.PythonVersion, pythonVersion) {
+						ensembler = config.EnsemblerData{
+							PythonVersion: strings.ReplaceAll(e.PythonVersion, ".", "-"),
+							EnsemblerID:   e.EnsemblerID,
+						}
+						break
+					}
+				}
+				Expect(ensembler.EnsemblerID).ShouldNot(Equal(0))
+			}
+
+			testData := TestData{cfg, ensembler}
+			router = e.POST("/projects/{projectID}/routers", cfg.Project.ID).
+				WithJSON(JSONPayload(payloadTpl, testData)).
+				Expect().Status(http.StatusOK).
+				JSON().Object()
+
+			DeferCleanup(func() {
+				Expect(
+					cluster.CleanupRouterDeployment(cfg.Project.Name, router.Value("name").String().Raw()),
+				).To(Succeed())
+			})
+
+			routerID := int(router.Value("id").Number().Raw())
+
+			Eventually(func(g Gomega) {
+				router = api.GetRouter(e, cfg.Project.ID, routerID)
+
+				g.Expect(router.Value("status").Raw()).ShouldNot(Equal(api.Status.Pending))
+			}, defaultDeploymentIntervals...).Should(Succeed())
+
+			router.
+				ValueEqual("status", "deployed").
+				Value("config").Object().ValueEqual("version", 1)
+
+			endpoint, err := url.Parse(router.Value("endpoint").String().Raw())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			endpoint.Path = "/"
+
+			routerCtx = RouterContext{
+				ID:        routerID,
+				ProjectID: cfg.Project.ID,
+				Endpoint:  endpoint.String(),
+			}
+
+		})
+
+		When(fmt.Sprintf("%s router is deployed", testCaseName), func() {
+			// Istio VirtualService configuration is applied asynchronously, so the fact that it
+			// exists in the cluster doesn't mean that it was already being in use.
+			// In the newer version of Istio (starting from v1.6), it is possible to wait for the
+			// changes readiness: https://istio.io/latest/docs/ops/configuration/mesh/config-resource-ready/
+			// However, we are currently on Istio v1.3 and there is no straight-forward
+			// way of achieving this readiness check, hence we just add an arbitrary sleep to give
+			// Istio some time to apply the changes.
+			//
+			// TODO: Remove once Turing is migrated on a newer version of Istio
```

Comment:
Just noticed this TODO - something to address, but we could do it in a separate MR later.